### PR TITLE
GitHub Issue NOAA-EMC/GSI#118.  EFSOI additions to util/EFSOI-Utilities/src directory.

### PR DESCRIPTION
**Pull request by AndrewEichmann-NOAA**
Addition of util/EFSOI-Utilities/src  directory and contents: EFSOI-specific fortran code and cmake files - same content as PR 124, which was to be merged from forked master branch, which was then reverted so that the changes would be in this PR-specific branch. PR 124 will be closed. 

CMakeLists.txt
efsoi.f90
efsoi_main.f90
gridio_efsoi.f90
loadbal_efsoi.f90
loc_advection.f90
scatter_chunks_efsoi.f90
statevec_efsoi.f90

### CODE REVIEW

**Comment by AndrewEichmann-NOAA** [comment]
@CatherineThomas-NOAA @aerorahul @MichaelLueken-NOAA 
This PR incorporates the changes previous applied to PR124 (EFSOI specific source files in util/EFSOI-Utilities/src )

File: util/EFSOI_Utilities/src/gridio_efsoi.f90 [hunk 1]
```diff
@@ -0,0 +1,898 @@
...
+    deallocate(ak,bk)
+	
+ endif
+ 
+!!! LL
+ if (use_gfs_nemsio) call nemsio_close(gfile,iret=iret)
+ if (use_gfs_ncio)   call close_dataset(dset)   
+  
+  
+
```
Comment (CatherineThomas-NOAA) [inline util/EFSOI_Utilities/src/gridio_efsoi.f90:350]: @AndrewEichmann-NOAA Can you easily remove these excess empty spaces?

File: util/EFSOI_Utilities/src/gridio_efsoi.f90 [hunk 2]
```diff
@@ -0,0 +1,898 @@
...
+  ! Transformation to EFSOI relevant quantities
+  ! Surface pressure contribution
+  grdin(:,levels(n3d) + ps_ind) = rdtrpr * grweight(:) * 100._r_kind * grdin(:,levels(n3d) + ps_ind)
+
+
+
+  
+  deallocate(psg)
+ 
+  ! updated by LL
```
Comment (CatherineThomas-NOAA) [inline util/EFSOI_Utilities/src/gridio_efsoi.f90:877]: @AndrewEichmann-NOAA Comments with author initials should not be included in the code itself (there are a few instances in this code).  Developer acknowledgement should only be in the top documentation block.

File: util/EFSOI_Utilities/src/scatter_chunks_efsoi.f90 [hunk 3]
```diff
@@ -0,0 +1,156 @@
+module scatter_chunks_efsoi
+
+use mpisetup !, only: numproc, nproc, mpi_real4
```
Comment (CatherineThomas-NOAA) [inline util/EFSOI_Utilities/src/scatter_chunks_efsoi.f90:3]: @AndrewEichmann-NOAA  Another instance of blanket "use" statements that is undesirable

File: util/EFSOI_Utilities/src/statevec_efsoi.f90 [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+!               io functionality necessary for efsoi calculations
+!   2021-03-04  Eichmann: updated to work with FV3 GFS
+!   2021-03-15  Eichmann: eliminated gridinfo_efsoi for src/enkf version 
+!
+! attributes:
+!   language: f95
+!
+!$$$
+
+use mpisetup
```
Comment (CatherineThomas-NOAA) [inline util/EFSOI_Utilities/src/statevec_efsoi.f90:54]: @AndrewEichmann-NOAA Another one

File: util/EFSOI_Utilities/src/statevec_efsoi.f90 [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+do k=1,nlevs
+   id_u(k) = clevels(u_ind-1) + k
+   id_v(k) = clevels(v_ind-1) + k
+   id_t(k) = clevels(tv_ind-1) + k
+   id_q(k) = clevels(q_ind-1) + k
+end do
+id_ps = clevels(nc3d) + ps_ind
+
+
+
```
Comment (CatherineThomas-NOAA) [inline util/EFSOI_Utilities/src/statevec_efsoi.f90:217]: @AndrewEichmann-NOAA Another set of blank lines

**Review by CatherineThomas-NOAA — APPROVED** [review]
Nothing major on my end.  Once the minor comments I left are addressed, consider this approved.

File: util/EFSOI_Utilities/src/gridio_efsoi.f90 (hunk 1 above)
Comment (AndrewEichmann-NOAA) [reply util/EFSOI_Utilities/src/gridio_efsoi.f90:350]: done

File: util/EFSOI_Utilities/src/gridio_efsoi.f90 (hunk 2 above)
Comment (AndrewEichmann-NOAA) [reply util/EFSOI_Utilities/src/gridio_efsoi.f90:877]: done, removed all in file, found no more instances of LL in this PR

File: util/EFSOI_Utilities/src/scatter_chunks_efsoi.f90 (hunk 3 above)
Comment (AndrewEichmann-NOAA) [reply util/EFSOI_Utilities/src/scatter_chunks_efsoi.f90:3]: fixed:
-use mpisetup !, only: numproc, nproc, mpi_real4
-use kinds, only: i_kind, r_single
+use mpisetup, only: numproc, nproc, mpi_real4
+use mpimod, only : mpi_comm_world
+use kinds, only: i_kind, r_single, r_kind

File: util/EFSOI_Utilities/src/statevec_efsoi.f90 (hunk 4 above)
Comment (AndrewEichmann-NOAA) [reply util/EFSOI_Utilities/src/statevec_efsoi.f90:54]: fixed:
-use mpisetup
+use mpisetup, only: numproc,nproc,mpi_wtime
+use mpimod, only: mpi_comm_world

File: util/EFSOI_Utilities/src/statevec_efsoi.f90 (hunk 5 above)
Comment (AndrewEichmann-NOAA) [reply util/EFSOI_Utilities/src/statevec_efsoi.f90:217]: removed

**Comment by AndrewEichmann-NOAA** [comment]
@CatherineThomas-NOAA Branch updated with suggested modifications

**Comment by CatherineThomas-NOAA** [comment]
@AndrewEichmann-NOAA Great, thanks for addressing my comments.

**Comment by MichaelLueken-NOAA** [comment]
@AndrewEichmann-NOAA, Since @CatherineThomas-NOAA has approved your changes, I will now work to merge the changes in this PR into the authoritative master.  However, similar to the previous PRs, you will need to reduce the number of commits to one:

1) Clone a temporary update copy of your fork
    git clone --recursive git@github.com:AndrewEichmann-NOAA/GSI.git update

2) Checkout your EXP-efso_fv3_PR_util_EFSOI_src branch
    git checkout EXP-efso_fv3_PR_util_EFSOI_src

3) Update your fix submodule to match that of your branch
    git submodule update

4) Rebase the NOAA-EMC/GSI master into your PR branch (Please note that this process will be trickier than the previous PRs)
```
    git remote add upstream https://github.com/NOAA-EMC/GSI.git
    git remote update
    git rebase upstream/master
    git add fix
    git submodule update
    git rebase --continue
    vi CMakeLists.txt
    remove lines 263 and 276-287
    git add CMakeLists.txt
    vi jobs/JGDAS_EFSOI
    remove lines 109-120
    git add jobs/JGDAS_EFSOI
    vi jobs/JGDAS_EFSOI_FCST
    remove lines 119-129
    git add jobs/JGDAS_EFSOI_FCST
    vi util/EFSOI_Utilities/scripts/README_fv3gfs
    remove lines 23-37
    git add util/EFSOI_Utilities/scripts/README_fv3gfs
    git rebase --continue
    vi CMakeLists.txt
    remove lines 263 and 275-282
    git add CMakeLists.txt
    git add jobs/JGDAS_EFSOI jobs/JGDAS_EFSOI_FCST
    git add util/EFSOI_Utilities/scripts/README_fv3gfs
    git rebase --continue
    git submodule update
```

5) Clean up commit messages
```
    git reset --soft HEAD~1
    git reset --soft HEAD~1
    git reset --soft HEAD~1
    git reset --soft HEAD~1
    git reset --soft HEAD~1
    git reset --soft HEAD~1
    git reset --soft HEAD~1
    git reset --soft HEAD~1
    git reset --soft HEAD~1
    git reset --soft HEAD~1
    git reset --soft HEAD~1
    git reset HEAD util/EFSOI_Utilities/scripts/convdata_codes.csv util/EFSOI_Utilities/scripts/display_osense.py util/EFSOI_Utilities/scripts/efso_advance_ensemble.sh util/EFSOI_Utilities/scripts/efso_cleanup.sh util/EFSOI_Utilities/scripts/efso_cleanup_wrapper.sh util/EFSOI_Utilities/scripts/efso_file_assembly.sh util/EFSOI_Utilities/scripts/efso_run.sh util/EFSOI_Utilities/scripts/efso_wrapper.sh util/EFSOI_Utilities/scripts/osense.py util/EFSOI_Utilities/scripts/read_osense.py
    git checkout -- util/EFSOI_Utilities/scripts/convdata_codes.csv util/EFSOI_Utilities/scripts/display_osense.py util/EFSOI_Utilities/scripts/osense.py util/EFSOI_Utilities/scripts/read_osense.py
    rm -f util/EFSOI_Utilities/scripts/efso_advance_ensemble.sh util/EFSOI_Utilities/scripts/efso_cleanup.sh util/EFSOI_Utilities/scripts/efso_cleanup_wrapper.sh util/EFSOI_Utilities/scripts/efso_file_assembly.sh util/EFSOI_Utilities/scripts/efso_run.sh util/EFSOI_Utilities/scripts/efso_wrapper.sh
    git reset HEAD jobs/JGDAS_EFSOI_ECEN jobs/JGDAS_EFSOI_POST jobs/JGDAS_EFSOI_SFC jobs/JGDAS_EFSOI_UPDATE
    git checkout -- jobs/JGDAS_EFSOI_ECEN jobs/JGDAS_EFSOI_POST jobs/JGDAS_EFSOI_SFC jobs/JGDAS_EFSOI_UPDATE
    cd fix
    git checkout rev2
    cd ..
    git add fix
```

6) Commit and push back to the repository
    git commit -m "GitHub Issue NOAA-EMC/GSI#118. Implement EFSOI process in FV3 GFS workflow - EFSOI-specific util source code."
    git push origin EXP-efso_fv3_PR_util_EFSOI_src --force

While Hera is updating the firewall to allow the access to VLab's new IP, Mars currently works.  Please let me know if you encounter any issues.

Mike

**Comment by AndrewEichmann-NOAA** [comment]
@MichaelLueken-NOAA Thanks Mike. I don't currently have access to Mars, but once the vlab issue gets cleared I will do these steps and let you know when I'm done.

**Comment by AndrewEichmann-NOAA** [comment]
@MichaelLueken-NOAA Ok, done. Are there some practices I am missing to make this task easier for you?

**Comment by MichaelLueken-NOAA** [comment]
@AndrewEichmann-NOAA It appears that you missed the last step:

git add fix

before committing and pushing.  Please use:

1) git add fix
2) git commit --amend (save and close without updating commit message)
3) git push origin EXP-efso_fv3_PR_util_EFSOI_src  --force

In the future, when you merge work into a branch to create a new PR (whether the branch is master or a new branch off of master), please make sure that you do a squash merge:

git merge --squash

this will reduce the merges in the branch to a single commit that you can set using:

git commit -m "GitHub Issue NOAA-EMC/GSI#XXXX. Insert message here."

When reviewers ask you to make changes, you would:

1) Make requested changes
2) git add _updated files_
3) git commit --amend
4) Close out of the commit window without changing the commit message
5) git push origin _branch name_ --force

This will allow you to make requested changes without adding additional commits to your PR.

**Comment by AndrewEichmann-NOAA** [comment]
@MichaelLueken-NOAA Sorry, did the git add fix just now. Output below:

(base) -bash-4.2$ git add fix
(base) -bash-4.2$ git commit --amend 
[EXP-efso_fv3_PR_util_EFSOI_src eb04fbcba] modification requested in review of PR 155
 Date: Mon May 24 17:20:26 2021 +0000
 3 files changed, 9 insertions(+), 33 deletions(-)
(base) -bash-4.2$ git push origin EXP-efso_fv3_PR_util_EFSOI_src --force
Enumerating objects: 29, done.
Counting objects: 100% (29/29), done.
Delta compression using up to 80 threads.
Compressing objects: 100% (25/25), done.
Writing objects: 100% (26/26), 24.21 KiB | 4.03 MiB/s, done.
Total 26 (delta 10), reused 1 (delta 0)
remote: Resolving deltas: 100% (10/10), completed with 2 local objects.
To github.com:AndrewEichmann-NOAA/GSI.git
 + 3c67ace38...eb04fbcba EXP-efso_fv3_PR_util_EFSOI_src -> EXP-efso_fv3_PR_util_EFSOI_src (forced update)
(base) -bash-4.2$

**Comment by MichaelLueken-NOAA** [comment]
@CatherineThomas-NOAA has approved this update and since no source code modifications has been made, I will now give final approval and merge this work to the authoritative repo.

**Pull request merged**